src/server_manager/ssm.py

```diff
@@ -0,0 +1,102 @@
+from __future__ import annotations
+import typing
+from typing import Optional
+from models.connection import ConnectionModel
+from models.wg_server import WgServerModel
+from vpn_manager.peers import Peer
+import logging
+from server_manager import ConnectionException, extract_wg_server_config
+from server_manager.interface import AbstractServerManager
+import time
+from typing import List, Union
+
+import boto3
+from botocore.config import Config
+from botocore.exceptions import ClientError, NoCredentialsError, PartialCredentialsError
+
+if typing.TYPE_CHECKING:
+    from vpn_manager.vpn import VpnServer
+
+
+log = logging.getLogger(__name__)
+
+
+class SsmConnection(AbstractServerManager):
+    @staticmethod
+    def _remote_ssm_command(cmd: str, connection_info: ConnectionModel) -> Union[List[str], str]:
+        """
+        Run a single shell command on an EC2 instance via SSM.
+        Returns stdout as a list of lines on success, or stderr as a string on error.
+        """
+        # build AWS credentials and client
+        try:
+            if connection_info.data.aws_access_key_id and connection_info.data.aws_secret_access_key:
+                ssm = boto3.client(
+                    "ssm",
+                    region_name=connection_info.data.region,
+                    aws_access_key_id=connection_info.data.aws_access_key_id,
+                    aws_secret_access_key=connection_info.data.aws_secret_access_key,
+                    config=Config(retries={"max_attempts": 3, "mode": "standard"}),
+                )
+            else:
+                # revert to the IAM role of the instance if no credentials are provided
+                ssm = boto3.client(
+                    "ssm",
+                    region_name=connection_info.data.region,
+                    config=Config(retries={"max_attempts": 3, "mode": "standard"}),
+                )
+            resp = ssm.send_command(
+                InstanceIds=[connection_info.data.target_id],
+                DocumentName="AWS-RunShellScript",
+                Parameters={"commands": [cmd]},
+                TimeoutSeconds=600,
+            )
+            cmd_id = resp["Command"]["CommandId"]
+        except ClientError as e:
+            return f"SSM connection failed: {e}"
+
+        # poll until the command finishes
+        while True:
+            try:
+                inv = ssm.get_command_invocation(
+                    CommandId=cmd_id, InstanceId=connection_info.data.target_id, PluginName="aws:RunShellScript"
+                )
+            except ClientError as e:
+                return f"SSM get_command_invocation failed: {e}"
+
+            status = inv["Status"]
+            if status in ("Success", "Failed", "Cancelled", "TimedOut"):
+                break
```

Comment:
This will work around a potential race condition where the command invocation isn't actually available on the server yet.  If you wait a second it'll work fine.
```suggestion
            try:
                inv = ssm.get_command_invocation(
                    CommandId=cmd_id, InstanceId=connection_info.data.target_id, PluginName="aws:RunShellScript"
                )
                status = inv["Status"]
                if status in ("Success", "Failed", "Cancelled", "TimedOut"):
                    break
            except ClientError as e:
                if not e.response['Error']['Code'] == 'InvocationDoesNotExist':
                    return f"SSM get_command_invocation failed: {e}"
```